architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,641 @@
+
+- Resource providers (non-declarative) are deployed per-location.
+- New functionality (API versions + types) is rolled out per-location.
+
+Some examples:
+
+> Resource provider `Contoso.Example` is only available in the `westus-1` location. UCP must reject creation requests to other locations.
+
+> Resource type `Contoso.Example/widgets` is available in all locations, but new API version `2024-08-01-preview` is only in the `eastus-2` location. UCP should block all requests for unsupported API version x location pairs.
+
+The design proposal includes axes for configuring location affinity:
+
+- `System.Resources/resourceProviders/locations` allows separate configuration of each location with an optional `address` in each location for the resource provider.
+- `System.Resources/resourceProviders/locations` includes the list of supported resource types and supported API versions.
+- `System.Resources/resourceProviders/resourceTypes` is the canonical definition for a given resource type.
+- `System.Resources/resourceProviders/resourceTypes/apiVersions` is the canonical definition for a given API version of a resource type.
+
+The structure and semantics of each resource type and API version must have a single definition. Each location may specify the resource types and API version it supports, but may not provide different definitions or semantics for resource types and API versions.
+
+### Architecture Diagram
+
+Resource type registration and querying is implemented in UCP. Any client may call the registration APIs or query APIs including resource providers.
+
+<center>
+
+![architecture diagram](./2024-08-user-defined-types-registration/registration.png)
+
+**Fig: Diagram of resource type registration and querying** [link](https://excalidraw.com/#json=FzQQaxv1d-vlmmDBAeOfc,UUXIasjO8p27hN0AB8qeiw)
+
+</center>
+
+### Detailed Design
+
+#### Nested resource type primer 
+
+Since this is our first nested resource type, a basic primer on nested types is needed. A nested resource type exists *inside* another resource type. A resource type is that is not nested is *top-level*. These relationships are also sometimes referred to as a parent and child. 
+
+The following statements are all accurate:
+
+- The `System.Resources/resourceProviders` type is the parent of `System.Resources/resourceProviders/resourceTypes`. 
+- The `System.Resources/resourceProviders/resourceTypes` type is the child of `System.Resources/resourceProviders`.
+- The `System.Resources/resourceProviders/resourceTypes` type is nested within `System.Resources/resourceProviders`.
+- The `System.Resources/resourceProviders` type is a top-level resource type.
+
+Any resource type with two segments (separated by `/`) is a top-level resource type. Any resource type with 3 or more segments (separate by `/`) is a nested resource type.
```

Comment:
The first one (two segments) is always a resource provider, right?

Reply:
The first segment is the resource provider namespace.